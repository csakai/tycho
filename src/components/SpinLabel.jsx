import React from 'react';

export default class SpinLabel extends React.Component {

  /**
   * Generates an array of n divs with the provided modifier.
   *
   * @param {String} modifier - BEM class modifier
   * @param {Number} count - number of arrows to provide
   * @return {<div />[]} array of divs with the expected class
   */
  createArrowSet = (modifier, count) => {
		const base = 'spin__arrow';
		const arrows = [];

		for (let i = 0; i < count; i++) {
      arrows.push(
        <span
          className={`${base} ${base}--${modifier}`}
          key={i}>
        </span>
      );
		}

		return (
			<div className={`${base}-container--${modifier}`}>
				{arrows}
			</div>
		);
	}

  render() {
    const {count, show} = this.props;
    const modifier = show ? 'show' : 'hide';

    return (
      <div className={`spin-container spin-container--${modifier}`}>
        <div className="spin">
          {this.createArrowSet('left', count)}
          <div className="spin__label">Spin</div>
          {this.createArrowSet('right', count)}
        </div>
      </div>
    );
  }
}

