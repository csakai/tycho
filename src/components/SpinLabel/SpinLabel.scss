@import 'globals.scss';

.spin-container {
    display: flex;
    position: absolute;
    top: 0;
    left: 0;
    right: 0;
    bottom: 0;
    opacity: 0;
    transition: opacity 0.5s;
    pointer-events: none;

    &--hide {
        opacity: 0;
    }

    &--show {
        opacity: 1;
    }
}

$n: 4; // # of arrows on either side

.spin {
    color: #fff;
    padding: $base-spacing;
    display: flex;
    width: 300px;
    margin: auto;
    align-items: center;
    justify-content: center;

    &__label {
        width: 5em;
        height: 5em;
        margin: $base-spacing;
        box-sizing: border-box;
        border-radius: 50%;
        color: $brand-secondary;
        background-color: $brand-primary;
        text-shadow: 1px 1px #e8e8e8;
        text-align: center;
        vertical-align: center;
        padding: 0 ($base-spacing / $n);
        display: flex;
        align-items: center;
        justify-content: center;
    }

    &__arrow-container {
        flex: 1;
        &--left {
            display: flex;
            flex-direction: row-reverse;
        }
    }

    &__arrow {
        animation: blink-animation ($transition-speed * 2) ease-in-out infinite;
        padding: 0 ($base-spacing / $n);
        font-weight: bold;

        @for $i from $n through 1 {
            &:nth-of-type(#{$i}) {
                animation-delay: #{($i / $n) - 1}s;
            }
        }

        @for $i from ($n + 1) through ($n * 2) {
            &:nth-of-type(#{$i}) {
                animation-delay: #{($i / $n)  - 1}s;
            }
        }

        &--left {
            &:before {
                content: '\2329';
            }
        }

        &--right {
            &:before {
                content: '\232A';
            }
        }
    }
}

@keyframes blink-animation {
    0% {
        opacity: 0;
    }
    50% {
        opacity: 1;
    }
    100% {
        opacity: 0;
    }
}

