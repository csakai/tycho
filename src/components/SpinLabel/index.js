import './SpinLabel.css';

export {default} from './SpinLabel';
