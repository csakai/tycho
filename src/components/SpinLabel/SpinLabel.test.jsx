import React from 'react';
import toJson from 'enzyme-to-json';
import {shallow} from 'enzyme';
import SpinLabel from './SpinLabel';

describe('SpinLabel Component', () => {
    describe('render()', () => {    
        it('should render the spin label with the `hide` class modifier if the show prop is false', () => {
            const component = shallow(<SpinLabel show={false} count={4} />);
      
            expect(toJson(component)).toMatchSnapshot();
            expect(component.find('.spin-container').hasClass('spin-container--hide')).toBe(true);
        });

        it('should render the spin label with the `show` class modifier if the show prop is true', () => {
            const component = shallow(<SpinLabel show={true} count={4} />);
      
            expect(toJson(component)).toMatchSnapshot();
            expect(component.find('.spin-container').hasClass('spin-container--show')).toBe(true);
        });
    });
});
