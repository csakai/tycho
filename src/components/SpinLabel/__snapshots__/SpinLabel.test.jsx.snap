// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SpinLabel Component render() should render the spin label with the \`hide\` class modifier if the show prop is false 1`] = `
<div
  className="spin-container spin-container--hide"
>
  <div
    className="spin"
  >
    <div
      className="spin__arrow-container--left"
    >
      <span
        className="spin__arrow spin__arrow--left"
        key="0"
      />
      <span
        className="spin__arrow spin__arrow--left"
        key="1"
      />
      <span
        className="spin__arrow spin__arrow--left"
        key="2"
      />
      <span
        className="spin__arrow spin__arrow--left"
        key="3"
      />
    </div>
    <div
      className="spin__label"
    >
      Spin
    </div>
    <div
      className="spin__arrow-container--right"
    >
      <span
        className="spin__arrow spin__arrow--right"
        key="0"
      />
      <span
        className="spin__arrow spin__arrow--right"
        key="1"
      />
      <span
        className="spin__arrow spin__arrow--right"
        key="2"
      />
      <span
        className="spin__arrow spin__arrow--right"
        key="3"
      />
    </div>
  </div>
</div>
`;

exports[`SpinLabel Component render() should render the spin label with the \`show\` class modifier if the show prop is true 1`] = `
<div
  className="spin-container spin-container--show"
>
  <div
    className="spin"
  >
    <div
      className="spin__arrow-container--left"
    >
      <span
        className="spin__arrow spin__arrow--left"
        key="0"
      />
      <span
        className="spin__arrow spin__arrow--left"
        key="1"
      />
      <span
        className="spin__arrow spin__arrow--left"
        key="2"
      />
      <span
        className="spin__arrow spin__arrow--left"
        key="3"
      />
    </div>
    <div
      className="spin__label"
    >
      Spin
    </div>
    <div
      className="spin__arrow-container--right"
    >
      <span
        className="spin__arrow spin__arrow--right"
        key="0"
      />
      <span
        className="spin__arrow spin__arrow--right"
        key="1"
      />
      <span
        className="spin__arrow spin__arrow--right"
        key="2"
      />
      <span
        className="spin__arrow spin__arrow--right"
        key="3"
      />
    </div>
  </div>
</div>
`;
